### DISCUSSION TITLE
Практикуем классы и ООП

### DISCUSSION THREAD

**Discussion by pingvin1308** [cleannetcode/CsharpForBeginers#3] [Новички]
1) Попробуйте повторить или доделать реализацию работы с задачами. Основные возможности программы:
Добавлять, удалять, редактировать и выводить список задач. А также отмечать задачу как выполненную.
Старайтесь сильно не усложнять новыми возможностями. Главное, чтобы в программе были элементы ООП)

2) Попробуйте спроектировать классы для программы HuntTheWumpus. Описание программы тут https://youtu.be/IuRqF3i0jps?t=6815

Сюда можно будет оставлять свои ссылки на решенную домашку. А также тут будет обсуждение ваших домашних заданий. Хочется верить, что мы здесь будем активно помогать друг другу)

**Comment by pingvin1308** [comment]


**Reply by AlexGoolyuk** [reply]
клево сделал

**Comment by trulander** [comment]
Моя домашка по прошлым урокам с учетом устранения недочетов с реализацией классов, наследования и всего что прошли. https://github.com/trulander/CsharpEducation/tree/main/blackjack

**Reply by pingvin1308** [reply]
Хорошая реализация, у тебя весьма неплохо получается

**Reply by trulander** [reply]
Спасибо. Стараюсь. Но у меня не хватает опыта в проектировании структуры проекта. Хочу полноценно научиться писать программы используя по максимуму ООП. Без твоих комментариев и уточнений как должна строиться программа, я бы все так и реализовал в базовом классе программы, как по функциональному программированию.

**Comment by trulander** [comment]
У меня есть проблема с домашкой. Я создал абстрактный класс для общих полей и заготовок методов. Но при попытке добавить поле в этот класс с типом другого класса, у меня компилятор ругается. Если сделать этот класс не абстрактным, включая методы, то ошибка пропадает. 
пример
```
  public abstract class Unit
    {
        //public Map _map; /*  на это поле ругается компилятор CS0052	Inconsistent accessibility: field type 'Map' is less accessible than field 'Unit._map'*/
        public int PositionX { get; set; }
        public int PositionY { get; set; }
        public Unit(Map map)
        {
            _map = map;
        }
        public abstract void Spawn();
        public abstract void ToGo();
    }
```

Также есть вопрос по нейминку полей которые массивы, и применять ли к ним геттеры и сеттеры, я хочу чтобы эти поля были публичными
```
public int SizeY { get; set; }
public int[] BusyX;
```

**Reply by pingvin1308** [reply]
field type 'Map' is less accessible than field 'Unit._map'
Написано, что у класса Map модификатор доступа сроже, чем у поля _map.
Возможно у тебя класс Map написан без модификатора доступа. Тогда он будет internal, а internal < public

 И еще есть ты пишешь поля _map, то лучше делать их приватными.
 
public int[] BusyX; Модификатор доступа массив, как и то будет ли это свойство и нет, задается абсолютно также как и у любого другого типа данных. В данном случае ты все почти оформил как свойство, возможно оно и должно статься ствойством, но тогда следует добавить { get; set; }

**Reply by trulander** [reply]
Спасибо.

**Comment by coder1coder** [comment]
**Домашка HuntTheWumpus**
[https://github.com/coder1coder/HuntTheWumpus](https://github.com/coder1coder/HuntTheWumpus)

**Reply by pingvin1308** [reply]
Посмотрел и отписал комменты в коммитах)

**Reply by coder1coder** [reply]
> Посмотрел и отписал комменты в коммитах)

Благодарю, на почту нотифи пришли, не разобрался только как на гитхабе выйти на твои коменты, приходится с почты по ссылкам идти.

**Reply by coder1coder** [reply]
Времени практически нету, обновил репу, сделаешь ревью?

**Reply by pingvin1308** [reply]
Отличная работа, очень близка к тому, что обычно у меня получается xD

Немного комментариев:
Direction - данный енум можно вынести в отдельный файл)
метод public virtual void Collision(Unit unit) - можно сделать абстрактным, так как ты не делаешь базовой реализации)

Еще я вижу, что метод Next класса Random доставил тебе боль)))
Вся проблема в том, что Next(0, 2) будет считать рандом от 0 до 1, если целыми числами. Он почему то не учитывает верхнюю границу, так написано в доке. 

Ты еще немного использовал синтаксис, который я не объяснял, что неплохо. Просто знай, что занятия про var-ы и лямбда выражения еще будут)

**Comment by trulander** [comment]
Вот моя домашка https://github.com/trulander/CsharpEducation/tree/main/HuntTheWumpus
В игре осталось реализовать разве что мультиплеер, остальное уже сделано:-D. Сделал формальные ограничения, но сейчас можно сыграть против 6 вампусов на поле в 20 на 20.
Думал получится по быстрому написать игру, в итоге потратил 3 дня.

**Reply by pingvin1308** [reply]
Выглядит сложно) 
Хорошо постарался,  как доберусь до компьютера дам комментариев в коде. А пока если вдруг получится сделай краткое описание классов и методов, а то сходу не все понятно)

**Reply by trulander** [reply]
Добавил описаний)), надеюсь более менее корректно

**Reply by pingvin1308** [reply]
Проверил на компьютере, более менее разобрался в реализации. Хорошо получилось использовать все разобранные на занятиях конструкции. Также вижу, что ты с успехом нашел нужные тебе инструменты по работе с консолью и так далее. Это отлично! :)
Немного сильно связал компоненты игры. В частности все игровые объекты имеют доступ по ссылке к карте. В твоем случае это дало тебе удобства, но обычно это приводит к потери в гибкости.

Немного про наименования переменных, свойств и полей. 
если используется тип bool, то лучше писать название в стиле вопроса. К примеру вместо Alive -> IsAlive
Также Busy я бы переименовал в Room. Так как по сути у тебя это комната)
А по определенному виду маркера (символа), ты выполняешь нужные тебе проверки)

Рекомендую попробовать реализовать эту игру еще раз по новой через некоторое время. Через месяц или два. Таким образом ты сможешь отработать навыки проектирования. Думаю мы за это время очень много всего пройдем и ты увидишь разницу и некоторую свободу в разработке на C# :)

**Reply by trulander** [reply]
Спасибо.
Я не знаю как можно писать код менее связанный между собой). Про обьект карты ты прав, мне не хватало во всех компонентах доступа к этому обьекту, видимо нужно было написать еще контролл для работы с этим обьектом, хотя он бы делал все тоже самое что и сама карта))). У меня еще затык был с тем что обьекты создаются в разное время и по коду их создания не все они доступны друг другу, хотелось создать их последовательно где то в отдельном классе каком нибудь init и через него иметь доступ ко всем обьектам сразу.
Вообщем это все вопрос архитектуры проектирования, я хромаю на обе ноги при реализации программы.
Под конец написания игры, мне было сложно додумывать последние штрихи логики, не мог дотянуться к нужным мне данным из разных кусков кода, поэтому делал какуюто дичь в виде проброса метода из карты в контроллер игры через обьект игрового поля, я понимаю что это еще больше связывает код, но как иначе то писать).

По неймингу полей понял спасибо)), в любом случае еще много будет нюансов где нужно будет корректировать).

**Comment by Amindlog** [comment]
https://github.com/Amindlog/CsharpForBeginners 
Классы не проектировал

**Reply by pingvin1308** [reply]
TodoList: кажется смотрел, но напишу еще тут. Все ок, только можно убрать метод swap. Кажется он там затесался))
А остальное пока действительно без изменений. Будут правки и эксперименты в коде, пиши :)

**Comment by emedvedeva** [comment]
Вот мое домашнее задание, я переделала свою телефонную книгу по аналогии с ToDoList:
https://github.com/emedvedeva/CsharpForBeginners/tree/main/HomeTask5

**Reply by pingvin1308** [reply]
Круто!
Кроме косметики в коде, особо и нечего прокомментировать)
С уверенностью движемся дальше :)

**Comment by AlexGoolyuk** [comment]
Ежедневник
Категорически извиняюсь, у меня всю неделю свет чинили, пришлось немного воспользоваться своими знаниями. Надеюсь смысл задания я понял правильно и суть была в практики работы с классами.
https://github.com/AlexGoolyuk/HomeWorks

**Reply by pingvin1308** [reply]
Извиняться не нужно, все отлично! :)
Написал комментариев, можно продолжать работать и улучшать программу)

**Comment by blood-mif** [comment]
Домашка по ToDoList
https://github.com/blood-mif/CsharpForBeginners/tree/main/ToDoList

**Reply by Z1pP** [reply]
Чекнул твой код, кое-что подсмотрел для себя

**Reply by pingvin1308** [reply]
Хорошая работа, продолжай в том же духе! :)

**Comment by 8bit-com** [comment]
https://github.com/8bit-com/BlackJack.git моя реализация BlackJack

**Reply by pingvin1308** [reply]
Я не учил так выводить методы! :)
Лучше такое не использовать, так как это усложняет восприятие кода.
Также нужно явно писать модификаторы доступа,

**Comment by 8bit-com** [comment]
https://github.com/8bit-com/ToDoList.git ToDoList

**Reply by pingvin1308** [reply]
Кажется не очень подходящем названия Arr что-нибудь для методов в классе Issues.
Логика расширения массива должна быть инкапсулирована в классе Issues. Достаточно сделать там метод для добавления задачи и проверку, хватает ли места в массиве.
Интересно сделал меню вывода, молодец! :)

**Comment by FanTei** [comment]
Поздно пришел ,но надеюсь посмотришь)
BlackJack and List
https://github.com/FanTei/MyWork

**Reply by pingvin1308** [reply]
Не обнаружил реализации работы с массивом в программе про List персон)

**Reply by FanTei** [reply]
Не доделал)

**Reply by FanTei** [reply]
https://github.com/FanTei/MyWork
Исправил сделал как в уроке и блэкджек тоже в порядок привел)

**Comment by FanTei** [comment]
Hunt the wumpus
https://github.com/FanTei/MyWork/tree/Main/Hunt%20the%20Wumpus

**Reply by pingvin1308** [reply]
Сори за задержку. Вампуса проверил и написал тебе в issues задач для исправления. К слову сейчас клево будет поработать над кодом, так как прошло время и все те огрехи в понятности кода тебе будет заметны)

**Comment by EvgenQ** [comment]
https://github.com/EvgenQ/HuntTheWumpusGame
Сделал 95% осталось мелочи
https://github.com/EvgenQ/ToDoList
это вроде бы готово

**Reply by pingvin1308** [reply]
Лист проверил, в целом работа неплохая для новичка. Набросал тебе задачек для исправления кода. 
Дальше даю тебе выбор. Мы можем довести программу с листом до его более менее идеального состояния, либо посчитать домашку выполненной (скажем на 4). 

Вампуса проверю после того, как получу ответ про туду лист))

**Reply by EvgenQ** [reply]
Роман, Вампуса не смотрели еще?

**Reply by pingvin1308** [reply]
Пока все в процессе)

**Comment by IlyaEliseev** [comment]
Hunt the wumpus. Пока реализовано передвижение и немного взаимодействие игрока с окружением.
https://github.com/IlyaEliseev/CsharpForBeginners/tree/main/HuntTheWumpus

**Reply by pingvin1308** [reply]
А ты ещё продолжаешь работать над этой программой? :)

**Reply by IlyaEliseev** [reply]
Привет. Нет моей "выпускной работой" по основам будет shop. Если я вернусь к wumpus, то буду переписывать с нуля.

**Reply by pingvin1308** [reply]
Супер, тогда буду стараться помогать тебе двигаться в разработке shop ;)
